packages/flutter_tools/lib/src/ios/fallback_discovery.dart

```diff
@@ -0,0 +1,156 @@
+import '../protocol_discovery.dart';
+import '../reporting/reporting.dart';
+
+/// A protocol for discovery of a vmservice on an attached iOS device with
+/// multiple fallbacks.
+///
+/// On versions of iOS 13 and greater, libimobiledevice can no longer listen to
+/// logs directly. The only way to discover an active observatory is through the
+/// mDNS protocol. However, there are a number of circumstances where this breaks
+/// down, such as when the device is connected to certain wifi networks or with
+/// certain hotspot connections enabled.
+///
+/// Another approach to discover a vmservice is to attempt to assign a
+/// specific port and then attempt to connect. This may fail if the port is
+/// not available. This port value should be either random, or otherwise
+/// generated with application specific input. This reduces the chance of
+/// accidentally connecting to another running flutter application.
+///
+/// Finally, if neither of the above approaches works, we can still attempt
+/// to parse logs.
+///
+/// To improve the overall resilience of the process, this class combines the
+/// three discovery strategies. First it assigns a port and attempts to connect.
+/// Then if this fails it falls back to mDNS, then finally attempting to scan
+/// logs.
+class FallbackDiscovery {
+  FallbackDiscovery({
+    @required DevicePortForwarder portForwarder,
+    @required MDnsObservatoryDiscovery mDnsObservatoryDiscovery,
+    @required Logger logger,
+    @required ProtocolDiscovery protocolDiscovery,
+    Future<VmService> Function(String wsUri, {Log log}) vmServiceConnectUri = vm_service_io.vmServiceConnectUri,
+  }) : _logger = logger,
+       _mDnsObservatoryDiscovery = mDnsObservatoryDiscovery,
+       _portForwarder = portForwarder,
+       _protocolDiscovery = protocolDiscovery,
+       _vmServiceConnectUri = vmServiceConnectUri;
+
+  final DevicePortForwarder _portForwarder;
+  final MDnsObservatoryDiscovery _mDnsObservatoryDiscovery;
+  final Logger _logger;
+  final ProtocolDiscovery _protocolDiscovery;
+  final Future<VmService> Function(String wsUri, {Log log}) _vmServiceConnectUri;
+
+  /// Attempt to discover the observatory port.
+  Future<Uri> discover({
+    @required int assumedDevicePort,
+    @required String packageId,
+    @required Device deivce,
+    @required bool usesIpv6,
+    @required int hostVmservicePort,
+    @required String packageName,
+  }) async {
+    final Uri result = await _attemptServiceConnection(
+      assumedDevicePort: assumedDevicePort,
+      hostVmservicePort: hostVmservicePort,
+      packageName: packageName,
+    );
+    if (result != null) {
+      return result;
+    }
+
+    try {
+      final Uri result = await _mDnsObservatoryDiscovery.getObservatoryUri(
+        packageId,
+        deivce,
+        usesIpv6: usesIpv6,
+        hostVmservicePort: hostVmservicePort,
+      );
+      if (result != null) {
+        UsageEvent('ios-mdns', 'success').send();
```

Comment:
It'll make it a little easier to view this data in analytics if first argument for all of the events is the same. So maybe make this:
```
UsageEvent('ios-handshake', 'mdns-success').send();
```
and similarly below.

Reply:
Are you worried about changing the existing event names?

Reply:
No. We're the only ones looking at those events in GA, and they'll kind of mean something different now anyway since the handshake procedure is different, so may as well make the naming uniform.

Reply:
Done. I kept the first set of events as just success and failure, since I wasn't sure what else to prefix them with